### PR TITLE
Refactor for IndexCoord to support cloud

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #18642 
Signed-off-by: cai.zhang <cai.zhang@zilliz.com>
Co-authored-by: Zach41 <zongmei.zhang@zilliz.com>

### CODE REVIEW

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/proto/index_coord.proto [hunk 1]
```diff
@@ -7,17 +7,18 @@ option go_package = "github.com/milvus-io/milvus/internal/proto/indexpb";
 import "common.proto";
 import "internal.proto";
 import "milvus.proto";
-import "schema.proto";
+import "data_coord.proto";
```
Comment (czs007) [inline internal/proto/index_coord.proto:10]: remove this

File: internal/distributed/indexnode/client/client.go [hunk 2]
```diff
@@ -108,86 +108,100 @@ func (c *Client) GetComponentStates(ctx context.Context) (*internalpb.ComponentS
...
 	})
 	if err != nil || ret == nil {
 		return nil, err
 	}
+	return ret.(*commonpb.Status), err
+}
 
-	return ret.(*internalpb.ShowConfigurationsResponse), err
+// GetJobNum query the task info of the index task.
+func (c *Client) GetJobNum(ctx context.Context, req *indexpb.GetJobNumRequest) (*indexpb.GetJobNumResponse, error) {
```
Comment (czs007) [inline internal/distributed/indexnode/client/client.go:167]: better to rename as GetJobStats

File: internal/proto/index_coord.proto [hunk 3]
```diff
@@ -26,16 +26,49 @@ service IndexCoord {
...
+message IndexInfo {
+  int64 collectionID = 1;
+  int64 fieldID = 2;
+  string index_name = 3;
+  int64 indexID = 4;
+  repeated common.KeyValuePair type_params = 5;
+  repeated common.KeyValuePair index_params = 6;
+}
+
+message FieldIndex {
```
Comment (czs007) [inline internal/proto/index_coord.proto:49]: Where is this used?

File: internal/proto/index_node.proto [hunk 4]
```diff
@@ -0,0 +1,69 @@
+syntax = "proto3";
+//package milvus.proto.indexnode;
+//option go_package = "github.com/milvus-io/milvus/internal/proto/indexnodepb";
```
Comment (czs007) [inline internal/proto/index_node.proto:3]: all content is commented?

File: internal/proxy/impl.go [hunk 5]
```diff
@@ -2656,6 +2657,7 @@ func (node *Proxy) Search(ctx context.Context, request *milvuspb.SearchRequest)
 		sentSize := proto.Size(qt.result)
 		metrics.ProxyReadReqSendBytes.WithLabelValues(strconv.FormatInt(Params.ProxyCfg.GetNodeID(), 10)).Add(float64(sentSize))
 	}
+	log.Debug("search done", zap.Int64s("topks", qt.result.Results.Topks))
```
Comment (czs007) [inline internal/proxy/impl.go:2660]: remove this

**Comment by czs007** [comment]
delete the commented and unused code directly

File: internal/proto/query_coord.proto [hunk 6]
```diff
@@ -203,6 +203,7 @@ message SegmentLoadInfo {
 
 message FieldIndexInfo {
   int64 fieldID =1;
+  // deprecated
```
Comment (czs007) [inline internal/proto/query_coord.proto:206]: why deprecate this ?

Comment (czs007) [reply internal/proto/query_coord.proto:206]: `estimateSegmentSize` located in querycoord/util.go is still using this

Comment (czs007) [reply internal/proto/query_coord.proto:206]: The modification of queyrcoord.proto should be cautious, which may cause some incompatibilities.

File: internal/proto/index_coord.proto [hunk 7]
```diff
@@ -47,93 +80,166 @@ message RegisterNodeResponse {
...
+
+message DropIndexRequest {
+  int64 collectionID = 1;
+  string index_name = 3;
+  int64 fieldID = 2;
 }
 
-message GetIndexFilePathsRequest {
-  repeated int64 indexBuildIDs = 1;
+message DropCollectionRequest {
```
Comment (czs007) [inline internal/proto/index_coord.proto:154]: What's this?

File: internal/rootcoord/root_coord.go [hunk 8]
```diff
@@ -848,41 +668,40 @@ func (c *Core) SetIndexCoord(s types.IndexCoord) error {
...
+
+		return resp.GetStates(), nil
 	}
 
-	c.CallDropIndexService = func(ctx context.Context, indexID typeutil.UniqueID) (retErr error) {
+	c.CallDropCollectionIndexService = func(ctx context.Context, collID UniqueID) (retErr error) {
 		defer func() {
 			if err := recover(); err != nil {
-				retErr = fmt.Errorf("drop index from index service panic, msg = %v", err)
+				retErr = fmt.Errorf("build index panic, msg = %v", err)
```
Comment (czs007) [inline internal/rootcoord/root_coord.go:697]: wrong  log

File: internal/indexcoord/task.go [hunk 9]
```diff
@@ -73,11 +81,159 @@ func (bt *BaseTask) Notify(err error) {
...
+	index := &model.Index{
+		CollectionID: cit.req.CollectionID,
+		FieldID:      cit.req.FieldID,
+		IndexID:      cit.indexID,
+		IndexName:    cit.req.IndexName,
+		TypeParams:   cit.req.TypeParams,
+		IndexParams:  cit.req.IndexParams,
+		CreateTime:   cit.createTs,
+	}
+	cit.table.SetIndex(index)
```
Comment (czs007) [inline internal/indexcoord/task.go:154]: do not use SetIndex.
first createSegmentIndex
then createIndex
finally noitfy builder to build index.

File: internal/indexcoord/index_builder.go [hunk 10]
```diff
@@ -55,8 +55,8 @@ func newIndexBuilder(ctx context.Context, ic *IndexCoord, metaTable *metaTable,
 		meta:             metaTable,
 		ic:               ic,
 		tasks:            make(map[int64]indexTaskState, 1024),
-		notify:           make(chan struct{}, 1024),
-		scheduleDuration: time.Second * 1,
+		notify:           make(chan struct{}, 10),
```
Comment (czs007) [inline internal/indexcoord/index_builder.go:58]: change len of notify to 1 and use
select {
   notify <- struct {}{}
  default:
    
}
to notify

File: internal/indexcoord/util.go [hunk 11]
```diff
@@ -65,25 +63,25 @@ func estimateIndexSize(dim int64, numRows int64, dataType schemapb.DataType) (ui
...
-	if funcutil.SliceContain(vecDTypes, req.GetFieldSchema().GetDataType()) {
-		dim, err := getDimension(req)
-		if err != nil {
-			return 0, err
-		}
-		return estimateIndexSize(dim, req.GetNumRows(), req.GetFieldSchema().GetDataType())
-	}
-	return estimateScalarIndexSize(req)
-}
+//func estimateIndexSizeByReq(req *indexpb.BuildIndexRequest) (uint64, error) {
```
Comment (czs007) [inline internal/indexcoord/util.go:71]: just delete this

File: internal/indexcoord/util_test.go [hunk 12]
```diff
@@ -16,201 +16,192 @@
 
 package indexcoord
 
-import (
```
Comment (czs007) [inline internal/indexcoord/util_test.go:19]: delete this file

File: internal/metastore/kv/rootcoord/kv_catalog.go [hunk 13]
```diff
@@ -516,33 +586,33 @@ func (kc *Catalog) DropPartition(ctx context.Context, collectionID typeutil.Uniq
...
-		log.Error("drop partition update meta fail",
-			zap.Int64("collectionID", collectionInfo.CollectionID),
-			zap.Int64("indexID", dropIdxID),
-			zap.Error(err))
-		return err
-	}
-
-	return nil
-}
+//func (kc *Catalog) DropIndex(ctx context.Context, collectionInfo *model.Collection, dropIdxID typeutil.UniqueID) error {
```
Comment (czs007) [inline internal/metastore/kv/rootcoord/kv_catalog.go:589]: just delete this

File: internal/querycoord/global_meta_broker.go [hunk 14]
```diff
@@ -130,148 +128,148 @@ func (broker *globalMetaBroker) getRecoveryInfo(ctx context.Context, collectionI
...
-		return false, 0, err
-	}
-
-	if !response.EnableIndex {
-		log.Info("describe segment from rootCoord successfully",
-			zap.Int64("collectionID", collectionID),
-			zap.Int64("segmentID", segmentID),
-			zap.Bool("enableIndex", false))
-		return false, 0, nil
+//func (broker *globalMetaBroker) getIndexBuildID(ctx context.Context, collectionID UniqueID, segmentID UniqueID) (bool, int64, error) {
```
Comment (czs007) [inline internal/querycoord/global_meta_broker.go:131]: just delete this.

File: internal/indexcoord/index_coord.go [hunk 15]
```diff
@@ -232,7 +217,13 @@ func (i *IndexCoord) Init() error {
 		i.chunkManager = chunkManager
 
 		i.garbageCollector = newGarbageCollector(i.loopCtx, i.metaTable, i.chunkManager)
-		i.sched, err = NewTaskScheduler(i.loopCtx, i.idAllocator, i.chunkManager, i.metaTable)
+		i.flushedSegmentWatcher, err = newFlushSegmentWatcher(i.loopCtx, i.etcdKV, i.metaTable, i.indexBuilder, i)
```
Comment (czs007) [inline internal/indexcoord/index_coord.go:220]: better to not create a new one.
forget to call Stop of flushedSegmentWatcher

**Pull request merged**